syntax = "proto3";

package ommx;

// Decison variable which mathematical programming solver will optimize.enum
// Decision variables must have its kind, i.e. binary, integer, real or others and unique identifier of 64-bit integer.
// Decision variables may have its name and subscripts which are used to identify in modeling tools.
message DecisionVariable {
    // Kind of the decision variable
    enum Kind {
        Binary = 0;
        Integer = 1;
        Real = 2;
    }

    // Unique identifier of the decision variable
    uint64 id = 1;
    // Kind of the decision variable
    Kind kind = 2;

    // Human readable description of the decision variable.
    message Description {
        // Name of the decision variable.
        string name = 1;
        // The subscripts of a deicision variable which is defined as multi-dimensional array.
        // Empty list means that the decision variable is scalar
        repeated uint64 subscripts = 2;
    }

    // This is optional since the name and subscripts does not exist in general mathematical programming situation
    optional Description description = 3;
}
